### ISSUE TITLE
[Feature] CLI autocomplete

### ISSUE THREAD

**Issue by arcanis** [yarnpkg/berry#512]
**Describe the user story**

I'm a user and I'd like to type my commands faster by using my shell's auto-complete features.

**Describe the solution you'd like**

This is half a feature request for [Clipanion](https://github.com/arcanis/clipanion), the CLI framework that we use (and that I wrote) to power Yarn's CLI.

I think the best would be to extend Clipanion in order to support a new function called `omelette`, which would generate a tree similar to the one that [Omelette](https://github.com/f/omelette) consumes:

```js
omelette('hello').tree({
  how: {
    much: {
      is: {
        this: ['car'],
        that: ['house'],
      }
    }
    are: ['you'],
    many: ['cars', 'houses'],
  },
  where: {
    are: {
      you: ['from'],
      the: ['houses', 'cars'],
    },
    is: {
      // You can also add some logic with defining functions:
      your() {
        return ['house', 'car'];
      },
    }
  },
}).init()
```

Since Clipanion already operates with a kind of tree (which is then turned into a state machine), it should be fairly easy. The hardest part would likely be to find a way to power dynamic segments, such as package names. This is likely to be a followup task, though - having command names and options would already be a decent improvement.

**Describe the drawbacks of your solution**

It might prove more complex from an architectural standpoint than I expect.

**Describe alternatives you've considered**

We could let third-party write autocompletion scripts (which is what we've been doing for the duration of the v1), but now that we have a decent CLI framework that we control I think there's a lot we could do to improve the program UX. For example, typing `yarn add typescript@<tab>` and get offered the versions that `yarn add -i` would suggest.

**Comment by yamadayuki** [comment]
Hi 👋  If you don't mind, I would like to tackle this one.

> I'm a user and I'd like to type my commands faster by using my shell's auto-complete features.

I strongly agree with this!

**Comment by arcanis** [comment]
For sure! 😊 Does the suggested plan sound reasonable? Do you have any questions?

**Comment by yamadayuki** [comment]
> Does the suggested plan sound reasonable?

First, I plan to extend [`clipanion`](https://github.com/arcanis/clipanion) to support the autocompletion with [`omelette`](https://github.com/f/omelette). Enable to support the autocompletion function in the framework natively.
I consider providing the API to register positional argument completion candidates.

And then, to implement the autocompletion functionality with clipanion support.

> Do you have any questions?

If the `clipanion` support autocompletion using `omelette`, the cli built with `clipanion` has to consume some options such as `--completion, --compgen, ...`. So I suppose that the autocompletion is an opt-in function. Because the `clipanion` users might use `--completion` option in their command. How do you think about this?

**Comment by arcanis** [comment]
Perfect 👍

> So I suppose that the autocompletion is an opt-in function. Because the `clipanion` users might use `--completion` option in their command. How do you think about this?

That sounds about right. I think we can use a similar pattern to the one we've already used when exposing the Clipanion definitions: let the user define themselves a command that call an `omelette` function defined on the CLI class.

https://github.com/yarnpkg/berry/blob/master/packages/plugin-essentials/sources/commands/entries/clipanion.ts

**Comment by dsifford** [comment]
Hey @arcanis -- congrats with the recent stable release!

Maintainer of the [bash completions for yarn 1](https://github.com/dsifford/yarn-completion) here.

My only concern with using a javascript library for handling bash completions is that we'd end up in a situation similar to how npm currently is. That is, pressing tab results in a very slight, yet noticeable delay in actually receiving the completion.

Is there any way we might be able to salvage what I have for v1 and create a native bash completion script for v2?

**Comment by arcanis** [comment]
Hey @dsifford! The main potential issue I can foresee is plugins. Since they're able to add new commands, anything static will necessarily miss the plugins.

One potential solution would be to dynamically generate the bash completion bindings when a plugin is installed - for example we could have a `plugin-completion` which would implement a JS hook that would consume the Clipanion definitions and would generate a static `.yarn/completion.sh` file. The global completion binding would source this file, and things would stay fairly fast. I admit I'm not very well versed in the Bash completion syntax so let me know if what I say doesn't make sense!

**Comment by dsifford** [comment]
Ah, I see. Yes, that would be challenging to roll entirely into a native bash completion script for sure.

An alternative would be to provide native bash completions for the core commands and then in the native bash completions, fall back to ones generated on the fly for plugins.

Really, the one I care most about is the completions for `yarn run` as I use those dozens of times per day to complete script names from my package.json files. `npm run <tab>` is painful (actually, just trying it now it feels like it improved from how it was historically) whereas `yarn run <tab>` or `yarn s<tab>` (to complete `yarn some-script-name`) is instantaneous.

**Comment by cspotcode** [comment]
Is there any progress on this, or an issue on the clipanion repository I can follow?

**Comment by arcanis** [comment]
I believe @paul-soporan [took a shot at it](https://github.com/arcanis/clipanion/compare/master...paul-soporan:feat/completion#diff-06017e7083155290c9b984dcc96fb019afd60d59513dbcfa7447ba0b43195f9e) a few months ago, but it's on hold at the moment 🤔

**Comment by soredake** [comment]
Any progress on this?

**Comment by paul-soporan** [comment]
Yes, I have managed to implement completion in Clipanion on a wip branch and I'm aiming to finally open a PR this weekend: https://github.com/paul-soporan/clipanion/tree/completion

**Comment by chrisands** [comment]
@paul-soporan  are you planning add support to "workspace" and "run" autocompletions like [chrisands/zsh-yarn-completions](https://github.com/chrisands/zsh-yarn-completions)?

**Comment by paul-soporan** [comment]
Yes, completions will be handled by Clipanion's state machine and will be implemented for each `Command` class. This means that:
- we can leverage the entire power of the existing JS API to even provide completions for `yarn add` or `yarn config set`
- plugins can register commands with their own completions and Clipanion will handle everything
- command paths and option names (together with descriptions) will be completed out-of-the-box for anybody using Clipanion, as Clipanion already has the relevant information
- Forwarding completions (e.g. for `yarn workspace` / `yarn workspaces foreach`) will just be a matter of calling `cli.complete`. For example, the completion code for `yarn workspace` will be something similar to (the stringifier isn't implemented yet and I'll probably leave it out of the first iteration):
```ts
commandName = args = Option.String({
  completion: async ({ current, prefix }, command) => {
      const workspace: Workspace = getTheWorkspaceInstance(command.workspaceName);

      return command.cli.complete({ current: stringify(command, ['commandName', 'args'], { commandName: current }), prefix: stringify(command, ['commandName', 'args'], { commandName: prefix })}, {cwd: workspace.cwd});
  };
})

args = Option.Proxy({
  completion: async ({ current, prefix }, command) => {
      const workspace: Workspace = getTheWorkspaceInstance(command.workspaceName);

      return command.cli.complete({ current: stringify(command, ['commandName', 'args'], { args: current }), prefix: Option.stringify(command, ['commandName', 'args'], { args: prefix })}, {cwd: workspace.cwd});
  };
})
```

**Comment by SamB** [comment]
By the way, @paul-soporan's PR is at https://github.com/arcanis/clipanion/pull/89; it looks like they haven't had a chance to work on it for about a year and/or got writers' block? (Why is GitHub calling that 2 years?)

And just as a reminder, even after they get the completion functionality merged into CLIpanion, it'd still need to get wired up for Yarn; it sounds like Yarn would need to add *at least* some glue to let shell scripts access the completion stuff, and *someone* would probably have to write some shell script to hook that into bash et al? (It seemed like @paul-soporan was working on changes to Yarn alongside the CLIpanion changes, but they may not be ready for review at the same point, and they sounded somewhat involved, so review could potentially take a while as well.)
